File: Codechef/Nov/WEIRDO.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <algorithm>
#include <cmath>
#include <queue>
#include <cstring>
#include <sstream>
#include <map>
#include <climits>
#include <set>
#include <unordered_map>
#include <unordered_set>

using namespace std;

bool isVowel(char &ch){
	return (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u');
}

double powerZ(double x, int y){  
    double temp;  
    if(y == 0)  
        return 1;  
    temp = powerZ(x, y / 2);  
    if (y % 2 == 0)  
        return temp * temp;  
    else
    {  
        if(y > 0)  
            return x * temp * temp;  
        else
            return (temp * temp) / x;  
    }  
}  

bool isAZ(string &s){ 
	int n = s.size();
	int c = 0;
	for (int i=0; i<n; i++){
		bool check = isVowel(s[i]);
		c += check;
		if (!check && i-1 >= 0 && !isVowel(s[i-1])){
			return false;
		}
		if (!check && i-2 >= 0 && !isVowel(s[i-2])){
			return false;
		}
	}
	
	return (n+1)/2 <= c;
} 
int main(){
	freopen("input.txt", "r", stdin);
	int t;
	scanf("%d", &t);
	while (t--){
		int n, a = 0, b = 0;
		scanf("%d", &n);
		vector< double> mpA(26, 0);
		vector< double> mpB(26, 0);
		vector< double> mpAC(26, 0);
		vector< double> mpBC(26, 0);
		for (int i=0; i<n; i++){
			string s;
			cin>>s;
			vector< double> sp(26, 0);
			for (auto ch : s){
				sp[(int)ch - 97]++;
			}
			bool isA = isAZ(s);
			if (isA){
				a++;
				for (int i=0; i<26; i++){
					if (sp[i] > 0){
						mpAC[i]++;
						mpA[i] += sp[i];	
					}
				}
			}else{
				b++;
				for (int i=0; i<26; i++){
					if (sp[i] > 0){
						mpBC[i]++;
						mpB[i] += sp[i];	
					}
				}
			}
		}
		long double inf = 10000000;
		vector< long double > x, y;
		for (int i=0; i<26; i++){
			if (mpA[i] > 0){
				x.push_back(mpAC[i]);
				y.push_back(powerZ(mpA[i], a));
			}
			if (mpB[i] > 0){
				y.push_back(mpBC[i]);
				x.push_back(powerZ(mpB[i], b));
			}
		}

		long double ratio = 1;
		while (!x.empty() && !y.empty()){
			if (ratio < inf){
				ratio *= x.back(); x.pop_back();
			}else{
				ratio /= y.back(); y.pop_back();
			}
		}

		if (ratio > inf && y.empty()){
			
		}else{
			while (!y.empty()){
				ratio /= y.back(); y.pop_back();
			}
			while (!x.empty() and ratio < inf){
				ratio *= x.back(); x.pop_back();
			}
		}
		if (ratio > inf)
			puts("Infinity");
		else{
			printf("%.7Lf\n", ratio);			
		}

	}
 return 0;
}

File: Codechef/Nov/WEIRDO.py
from collections import defaultdict

def is_vowel(ch):
    return (ch == 'a' or ch == 'e' or ch == 'i' or ch == 'o' or ch == 'u')

def power(x, y): 
  
    if(y == 0): return 1
    temp = power(x, int(y / 2))  
      
    if (y % 2 == 0): 
        return temp * temp 
    else: 
        if(y > 0): return x * temp * temp 
        else: return (temp * temp) / x 

def checker(s):

    n = len(s)
    c = 0
    for i in range(0, n):
        check = is_vowel(s[i])
        c += check
        if ((not check) and i-1 >= 0 and (not is_vowel(s[i-1]))):
            return False
        
        if ((not check) and i-2 >= 0 and (not is_vowel(s[i-2]))):
            return False
        
    return int((n+1)/2) <= c

t = int(input())
i = 0
while i < t:
    n = int(input())
    a = 0
    b = 0
    mpA = [0.0]*26
    mpB = [0.0]*26
    mpAC = [0.0]*26
    mpBC = [0.0]*26

    k = 0
    while k < n:
        s = input()

        c = 0
        sp = [0]*26
        l = len(s)
        z = 0
        while z < l:
            sp[ord(s[z])-97] += 1
            z += 1
        is_a = checker(s)
        if is_a:
            a += 1
            x = 0
            while x < 26:
                if sp[x] > 0:
                    mpAC[x] += 1
                    mpA[x] += sp[x]
                x += 1
        else:
            b += 1
            x = 0
            while x < 26:
                if sp[x] > 0:
                    mpBC[x] += 1
                    mpB[x] += sp[x]
                x += 1
        k += 1
    A = 1.0
    B = 1.0

    x = 0
    while x < 26:
        if mpA[x] > 0:
            A *= mpAC[x]/power(mpA[x], a)
        
        if mpB[x] > 0:
            B *= mpBC[x]/power(mpB[x], b)

        x += 1

    if B == 0:
        print ("Infinity")
    else:
        ratio = A/B
        if ratio > 10000000:
            print ("Infinity")
        else:
            print ("{0:7f}".format(ratio))

    i += 1



